plugins/tpm-eventlog/fu-tpm-eventlog-device.c

```diff
@@ -0,0 +1,355 @@
+		if (!fu_common_read_uint32_safe	(buf, bufsz,
+						 idx + FU_TPM_EVENTLOG_V2_IDX_PCR,
+						 &pcr, G_LITTLE_ENDIAN, error))
+			return FALSE;
+		if (!fu_common_read_uint32_safe	(buf, bufsz,
+						 idx + FU_TPM_EVENTLOG_V2_IDX_TYPE,
+						 &event_type, G_LITTLE_ENDIAN, error))
+			return FALSE;
+		if (!fu_common_read_uint32_safe	(buf, bufsz,
+						 idx + FU_TPM_EVENTLOG_V2_IDX_DIGEST_COUNT,
+						 &digestcnt, G_LITTLE_ENDIAN, error))
+			return FALSE;
+
+		/* read checksum block */
+		idx += FU_TPM_EVENTLOG_V2_SIZE;
+		for (guint i = 0; i < digestcnt; i++) {
+			guint16 alg_type = 0;
+			guint32 alg_size = 0;
+			g_autofree gchar *csum_tmp = NULL;
+
+			/* get checksum type */
+			if (!fu_common_read_uint16_safe	(buf, bufsz, idx,
+							 &alg_type, G_LITTLE_ENDIAN, error))
+				return FALSE;
+			alg_size = fu_tpm_eventlog_hash_get_size (alg_type);
+			if (alg_size == 0) {
+				g_set_error (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_NOT_SUPPORTED,
+					     "hash algorithm 0x%x size not known",
+					     alg_type);
+				return FALSE;
+			}
+
+			/* build checksum */
+			idx += sizeof(alg_type);
+			csum_tmp = fu_common_read_strhex_safe (buf, bufsz, idx, alg_size, error);
+			if (csum_tmp == NULL)
+				return FALSE;
+
+			/* save this for analysis */
+			if (alg_type == TPM2_ALG_SHA1)
+				csum = g_steal_pointer (&csum_tmp);
+
+			/* next block */
+			idx += alg_size;
+		}
+
+		/* read data block */
+		if (!fu_common_read_uint32_safe	(buf, bufsz, idx,
+						 &datasz, G_LITTLE_ENDIAN, error))
+			return FALSE;
+
+		/* save blob if PCR=0 */
+		idx += sizeof(datasz);
+		if (pcr == ESYS_TR_PCR0) {
+			g_autofree guint8 *data = NULL;
+			FuTpmEventlogDeviceItem *item;
+
+			/* build item */
+			data = g_malloc0 (datasz);
+			if (!fu_memcpy_safe (data, datasz, 0x0,		/* dst */
+					     buf, bufsz, idx, datasz,	/* src */
+					     error))
+				return FALSE;
+
+			/* not normally required */
+			if (g_getenv ("FWUPD_TPM_EVENTLOG_VERBOSE") != NULL) {
+				fu_common_dump_full (G_LOG_DOMAIN, "Event Data",
+						     data, datasz, 20,
+						     FU_DUMP_FLAGS_SHOW_ASCII);
+			}
+			item = g_new0 (FuTpmEventlogDeviceItem, 1);
+			item->kind = event_type;
+			item->checksum = g_steal_pointer (&csum);
+			item->blob = g_bytes_new_take (g_steal_pointer (&data), datasz);
+			g_ptr_array_add (self->items, item);
+		}
+
+		/* next entry */
+		idx += datasz;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_tpm_eventlog_device_parse_blob (FuTpmEventlogDevice *self,
+				   const guint8 *buf, gsize bufsz,
+				   GError **error)
+{
+	gchar sig[] = FU_TPM_EVENTLOG_V2_HDR_SIGNATURE;
+
+	/* look for TCG v2 signature */
+	if (!fu_memcpy_safe ((guint8 *) sig, sizeof(sig), 0x0,		/* dst */
+			     buf, bufsz, FU_TPM_EVENTLOG_V1_SIZE,	/* src */
+			     sizeof(sig), error))
+		return FALSE;
+	if (g_strcmp0 (sig, FU_TPM_EVENTLOG_V2_HDR_SIGNATURE) == 0)
+		return fu_tpm_eventlog_device_parse_blob_v2 (self, buf, bufsz, error);
```

Comment:
It might be useful to note in a `g_debug` here that the blob was a v2 for future debugging points when somethign goes wrong.